### PR TITLE
WT-6295 Aggregate time windows in salvage

### PR DESCRIPTION
**Pull request by keithbostic**
@quchenhao, this branch does timestamp aggregation during salvage, would you please do the review?

I have a few concerns:

* First, in WT-6260, you commented "The cause of the failure is we are not aggregating time window information to the internal pages when doing salvage. If we need to aggregate that information, we have to rebuild every page we try to salvage, which will make salvage take much longer." I'm not following that, my approach is to read the timestamps out of the leaf pages and then use that information to initialize the internal pages we create. It seems to me you're talking about rewriting leaf pages, which would be expensive -- can you please explain further what you were thinking, and if I'm missing something?

* Second, the failure in the initial WT-6260 ticket was that a prepare state in a leaf page wasn't being propagated to the parent. The code I added will, in fact fix that by propagating the prepare up the tree, but I think that's wrong. I would expect rollback-to-stable to clear ANY prepare information on startup, but if we're salvaging the file, I suspect we can't depend on that, the file was presumably corrupted. Shouldn't we clear any prepare flags that are set in the salvaged fiile? Prepared operations can't possible be resolved after a salvage, so I think we should clear them, not propagate them up the tree. Thoughts?

* Finally, please check me on my timestamp macro use, specifically, I think I should be using `WT_TIME_AGGREGATE_INIT_MAX`, not `WT_TIME_AGGREGATE_INIT`, but I'm not confident in that decision.

### CODE REVIEW

**Comment by quchenhao** [comment]
> @quchenhao, this branch does timestamp aggregation during salvage, would you please do the review?
> 
> I have a few concerns:
> 
> * First, in WT-6260, you commented "The cause of the failure is we are not aggregating time window information to the internal pages when doing salvage. If we need to aggregate that information, we have to rebuild every page we try to salvage, which will make salvage take much longer." I'm not following that, my approach is to read the timestamps out of the leaf pages and then use that information to initialize the internal pages we create. It seems to me you're talking about rewriting leaf pages, which would be expensive -- can you please explain further what you were thinking, and if I'm missing something?
> * Second, the failure in the initial WT-6260 ticket was that a prepare state in a leaf page wasn't being propagated to the parent. The code I added will, in fact fix that by propagating the prepare up the tree, but I think that's wrong. I would expect rollback-to-stable to clear ANY prepare information on startup, but if we're salvaging the file, I suspect we can't depend on that, the file was presumably corrupted. Shouldn't we clear any prepare flags that are set in the salvaged fiile? Prepared operations can't possible be resolved after a salvage, so I think we should clear them, not propagate them up the tree. Thoughts?
> * Finally, please check me on my timestamp macro use, specifically, I think I should be using `WT_TIME_AGGREGATE_INIT_MAX`, not `WT_TIME_AGGREGATE_INIT`, but I'm not confident in that decision.

1. Yes, Keith. I don't know we can read the tw from the page itself. If that's the case, no need for us to rewrite the leaf pages.

2. If rollback to stable is not guaranteed to run after salvage, then I agree we should clear them. I made the change in WT-6238 to also write prepared updates to disk in salvage. But I think we will run rollback to stable anyway at shutdown and startup so this is fine?

File: test/format/salvage.c [hunk 1]
```diff
@@ -139,16 +139,14 @@ wts_salvage(void)
...
-    wts_verify("post-salvage verify");
-#endif
+    wts_verify(conn, "post-salvage verify");
     wts_close(&conn, &session);
 
     /* Corrupt the file randomly, salvage, then verify. */
     if (corrupt()) {
         wts_open(g.home, &conn, &session, false);
         testutil_check(session->salvage(session, g.uri, "force=true"));
+        wts_verify(conn, "post-corrupt-salvage verify");
```
Comment (quchenhao) [inline test/format/salvage.c:149]: I think this is deliberately removed by @luke-pearson 's change. Can you confirm we can do this again?

File: src/btree/bt_slvg.c [hunk 2]
```diff
@@ -596,13 +603,17 @@ __slvg_trk_leaf(WT_SESSION_IMPL *session, const WT_PAGE_HEADER *dsk, uint8_t *ad
...
+
         /*
          * Column-store variable-length format: the start key can be taken from the block's header,
          * stop key requires walking the page.
          */
         stop_recno = dsk->recno;
         WT_CELL_FOREACH_KV (session, dsk, unpack) {
             stop_recno += __wt_cell_rle(&unpack);
+
+            WT_TIME_AGGREGATE_UPDATE(&trk->trk_ta, &unpack.tw);
```
Comment (quchenhao) [inline src/btree/bt_slvg.c:616]: Should we aggregate the tws here? If we rewrite a page, we will do a reconciliation and the time windows should be automatically aggregated there I think.

File: src/btree/bt_slvg.c [hunk 3]
```diff
@@ -596,13 +603,17 @@ __slvg_trk_leaf(WT_SESSION_IMPL *session, const WT_PAGE_HEADER *dsk, uint8_t *ad
           trk->col_stop);
         break;
     case WT_PAGE_COL_VAR:
+        WT_TIME_AGGREGATE_INIT_MAX(&trk->trk_ta);
```
Comment (quchenhao) [inline src/btree/bt_slvg.c:606]: I agree init to max is correct.

File: src/btree/bt_slvg.c (hunk 2 above)
Comment (quchenhao) [reply src/btree/bt_slvg.c:616]: I have a ditched pr https://github.com/wiredtiger/wiredtiger/pull/5703 which relies on reconciliation to aggregate the time windows.

File: test/format/salvage.c (hunk 1 above)
Comment (keithbostic) [reply test/format/salvage.c:149]: I think that you removed it in #5704, as part of the WT-6260 changes.

Regardless: if it doesn't work, we have a problem. The metadata file is salvaged in some cases, and we should not create a metadata file we can't verify, that's not a good plan.

File: src/btree/bt_slvg.c (hunk 2 above)
Comment (keithbostic) [reply src/btree/bt_slvg.c:616]: This isn't rewriting the page, this is applying the leaf-page record to the aggregate time window for the page.

File: test/format/salvage.c (hunk 1 above)
Comment (quchenhao) [reply test/format/salvage.c:149]: I only comment it out. @luke-pearson removed it later in WT-6278

File: src/btree/bt_slvg.c (hunk 2 above)
Comment (quchenhao) [reply src/btree/bt_slvg.c:616]: oh, I see.

Comment (keithbostic) [reply src/btree/bt_slvg.c:609]: I think this is a good point, @quchenhao.

I added calls to `WT_TIME_AGGREGATE_MERGE` in the two routines that build internal pages (`__slvg_row_build_internal` and `__slvg_col_build_internal`), and I don't think they're necessary. The reconciliation code should do all of that work for me.

The same is true for the `WT_CKPT.write_gen` field, updated with the timestamp information in `__wt_checkpoint_tree_reconcile_update()`.

I've pushed that change in ee8dbd5a7.

**Comment by keithbostic** [comment]
@quchenhao 
> I think we will run rollback to stable anyway at shutdown and startup so this is fine? Also we can clear the prepared updates in the page we rewrite, but we can't delete the ones in the page we don't touch so I think we still should propagate them and ensure we run rollback to stable after that.

That makes sense to me, and if we get the prepare aggregation correct, rollback-to-stable will clean up for us.

The alternative would be to rewrite leaf pages that contain prepared items. (That's not hard to do, but I agree there's no point in doing the work if rollback-to-stable owns the problem.)

So, is the plan every WiredTiger run will include a rollback-to-stable at startup and shutdown? (Even when WiredTiger is run standalone?)

@agorrod, can you please comment?

File: test/format/salvage.c (hunk 1 above)
Comment (luke-pearson) [reply test/format/salvage.c:149]: I removed this line in WT-6266 as we currently are working under the assumption that if we intentionally corrupt a datafile, it is extremely unlikely that verify will succeed due to the issues discussed in that ticket. Please don't add this line back until we resolve the discussion in WT-6266.

**Comment by agorrod** [comment]
> So, is the plan every WiredTiger run will include a rollback-to-stable at startup and shutdown? (Even when WiredTiger is run standalone?)

Yes - since we are creating checkpoints where the specific time point needed can be spread across the history store and data files, it is always necessary to run rollback-to-stable on startup to ensure the data files are starting at the consistent point in time expected.
Note: we try to run rollback-to-stable on shutdown, so there isn't any work to be done on startup, but that's not guaranteed to be run, so we always need to at least check to see whether there is work to do on startup.

File: test/format/salvage.c (hunk 1 above)
Comment (keithbostic) [reply test/format/salvage.c:149]: @luke-pearson 
> I removed this line in WT-6266 as we currently are working under the assumption that if we intentionally corrupt a datafile, it is extremely unlikely that verify will succeed due to the issues discussed in that ticket. 

@luke-pearson, as I understand it, WT-6266 is discussing metadata verification, where verifying the HS file causes retrieval of records that may not exist in the underlying file. I believe the line of code we're discussing verifies the individual file after it has been salvaged, and so should not have problems. Or am I misunderstanding something?

Comment (luke-pearson) [reply test/format/salvage.c:149]: Alright my mistake, certainly __wt_verify verifies the specific tree and doesn't perform the backwards traversal from the history store to the data store file. It still does verify the history store for each key in the data store file but that doesn't cause the issue I saw.

**Pull request merged**